ls1intum/ArTEMiS
# Feature/Offline Alert

### Checklist

### Motivation and Context
When the user gets disconnected from the websocket, Artemis can appear to not be responding for any reason. It would be good to provide more information about this situation to the user.

### Description
Implemented new notification type: Connection Notification.

State Connected:
- Don't show notification

State Disconnected:
- Show persisting error notification

State Reconnected:
- Show temporary success notification
- After 10 seconds remove notification and go to state Connected

When testing this feature, we discovered that the disconnect callback was not triggered on chrome when there is no internet connection.
I found out that currently we don't send a heartbeat between server and client: https://docs.spring.io/spring/docs/current/spring-framework-reference/web.html#websocket
For some reason Firefox and Safari still receive a disconnect event from some other source, but Chrome doesn't. I configured the heartbeat and now it also works on Chrome.

Settings for the heartbeat are 10 seconds (outgoing), 20 seconds (incoming).
So we send a message every 10 seconds from client and expect to receive a message every 20 seconds.

Here is how the heartbeats are handled in webstomp client:
https://github.com/JSteunou/webstomp-client/blob/master/src/client.js#L347-L367

### Steps for Testing
1. Login to Artemis
2. Disable internet connection
=> Error alert should appear
3. Enable internet connection
=> Success alert should appear

### Screenshots

![image](https://user-images.githubusercontent.com/28230611/56977787-4a65a880-6b76-11e9-8b7a-53858ae4d588.png)

![image](https://user-images.githubusercontent.com/28230611/56978250-60279d80-6b77-11e9-84fb-8dceaf318b56.png)
